services/venice-controller/src/main/java/com/linkedin/venice/controller/DeferredVersionSwapService.java

```diff
@@ -0,0 +1,147 @@
+ * The wait time is specified through a store/version level config (target_swap_region_wait_time) and the default wait time is 60m.
+ */
+public class DeferredVersionSwapService extends AbstractVeniceService {
+  private final AtomicBoolean stop = new AtomicBoolean(false);
+  private final Set<String> allClusters;
+  private final VeniceControllerMultiClusterConfig veniceControllerMultiClusterConfig;
+  private final Admin admin;
+  private final Thread deferredVersionSwapThread;
+  private final Time time;
+  private final long sleepInterval;
+  private static final Logger LOGGER = LogManager.getLogger(DeferredVersionSwapService.class);
+
+  public DeferredVersionSwapService(Admin veniceHelixAdmin, VeniceControllerMultiClusterConfig multiClusterConfig) {
+    this(veniceHelixAdmin, multiClusterConfig, new SystemTime());
+  }
+
+  protected DeferredVersionSwapService(Admin admin, VeniceControllerMultiClusterConfig multiClusterConfig, Time time) {
+    this.admin = admin;
+    this.time = time;
+    this.allClusters = multiClusterConfig.getClusters();
+    this.deferredVersionSwapThread = new Thread(new DeferredVersionSwapTask(), "DeferredVersionSwapTask");
+    this.veniceControllerMultiClusterConfig = multiClusterConfig;
+    this.sleepInterval = multiClusterConfig.getDeferredVersionSwapSleepMs();
+  }
+
+  @Override
+  public boolean startInner() throws Exception {
+    deferredVersionSwapThread.start();
+    return true;
+  }
+
+  @Override
+  public void stopInner() throws Exception {
+    stop.set(true);
+    deferredVersionSwapThread.interrupt();
+  }
+
+  private Set<String> getRegionsForVersionSwap(Map<String, Integer> candidateRegions, Set<String> targetRegions) {
+    Set<String> remainingRegions = new HashSet<>(candidateRegions.keySet());
+    remainingRegions.removeAll(targetRegions);
+    return remainingRegions;
+  }
+
+  private class DeferredVersionSwapTask implements Runnable {
+    @Override
+    public void run() {
+      while (!stop.get()) {
+        try {
+          time.sleep(sleepInterval);
+        } catch (InterruptedException e) {
+          LOGGER.error("Received InterruptedException during sleep in DeferredVersionSwapTask thread");
+          break;
+        }
+
+        for (String cluster: allClusters) {
+          List<Store> stores = admin.getAllStores(cluster);
+          for (Store store: stores) {
+            if (StringUtils.isEmpty(store.getTargetSwapRegion())) {
+              continue;
+            }
+
+            Map<String, Integer> coloToVersions = admin.getCurrentVersionsForMultiColos(cluster, store.getName());
+            Set<String> targetRegions = RegionUtils.parseRegionsFilterList(store.getTargetSwapRegion());
+            Set<String> remainingRegions = getRegionsForVersionSwap(coloToVersions, targetRegions);
+
+            // Do not perform version swap for davinci stores
+            int targetVersion = store.getCurrentVersion();
```

Comment:
We are going to run this in parent controller right ? in parent controller current version for a store is not properly set. Get the version from target region of the store